File: android/AndroidPrinterAdapter.java
package com.nukkadshops.printer.android;

import android.content.Context;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AndroidPrinterAdapter {
    private Context context;
    private UsbAndroidTransport transport;
    private PrinterDiscovery discovery;
    private String connectedDeviceId;

    public AndroidPrinterAdapter(Context context) {
        this.context = context;
        this.transport = new UsbAndroidTransport(context);
        this.discovery = new PrinterDiscovery(context);
    }

    // Discover all USB printers
    public List<PrinterDiscovery.PrinterDevice> discoverPrinters() throws Exception {
        return discovery.discoverPrinters();
    }

    // Connect to a specific printer
    public boolean connect(String deviceId, int vendorId, int productId) throws Exception {
        boolean connected = transport.connect(deviceId, vendorId, productId);
        if (connected) {
            connectedDeviceId = deviceId;
        }
        return connected;
    }

    // Disconnect from printer
    public void disconnect() throws Exception {
        transport.disconnect();
        connectedDeviceId = null;
    }

    // Get printer status
    public Map<String, Object> getStatus() throws Exception {
        Map<String, Object> status = new HashMap<>();
        try {
            Object statusResult = transport.getStatus();
            status.put("status", statusResult);
            status.put("connected", connectedDeviceId != null);
            return status;
        } catch (Exception e) {
            status.put("error", e.getMessage());
            return status;
        }
    }

    // Print receipt
    public Map<String, Object> printReceipt(String jobId, String printerId, Map<String, Object> receiptPayload) throws Exception {
        Map<String, Object> result = new HashMap<>();
        try {
            EscPosReceiptRenderer renderer = new EscPosReceiptRenderer();
            byte[] receiptData = renderer.render(receiptPayload);

            boolean sent = transport.sendData(receiptData);
            if (sent) {
                result.put("status", "completed");
                result.put("message", "Receipt printed successfully");
                result.put("jobId", jobId);
            } else {
                result.put("status", "failed");
                result.put("message", "Failed to send data to printer");
            }
        } catch (Exception e) {
            result.put("status", "failed");
            result.put("message", "Error: " + e.getMessage());
        }
        return result;
    }

    // Print label
    public Map<String, Object> printLabel(String jobId, String printerId, Map<String, Object> labelPayload) throws Exception {
        Map<String, Object> result = new HashMap<>();
        try {
            SimpleLabelRenderer renderer = new SimpleLabelRenderer();
            byte[] labelData = renderer.render(labelPayload);

            boolean sent = transport.sendData(labelData);
            if (sent) {
                result.put("status", "completed");
                result.put("message", "Label printed successfully");
                result.put("jobId", jobId);
            } else {
                result.put("status", "failed");
                result.put("message", "Failed to send data to printer");
            }
        } catch (Exception e) {
            result.put("status", "failed");
            result.put("message", "Error: " + e.getMessage());
        }
        return result;
    }

    // Get capabilities
    public Map<String, Object> getCapabilities() throws Exception {
        Map<String, Object> capabilities = new HashMap<>();
        capabilities.put("printQuality", "MEDIUM");
        capabilities.put("maxPrintWidth", 384);
        capabilities.put("supportsReceipts", true);
        capabilities.put("supportsLabels", true);
        capabilities.put("protocols", new String[]{"ESC/POS", "TSC"});
        return capabilities;
    }
}

File: src/infrastructure/adapters/android/android-printer-adapter.ts
import type { PrinterAdapter } from "../../../domain/contracts/adapter";
import type { PrintJob, PrintResult } from "../../../domain/models/job";
import type { LabelPayload, ReceiptPayload } from "../../../domain/models/payloads";
import type {
  AdapterCapabilityQuery,
  PrinterCapability,
  PrinterDevice,
  PrinterStatus
} from "../../../domain/models/printer";
import { PrinterError } from "../../../domain/models/errors";
import {
  getAndroidNativePrinterModule,
  getFallbackAndroidPrinterCapabilities,
  type AndroidNativePrinterModule,
  type NativePrinterDevice
} from "../../../react-native/native-printer-module";
import { EscPosReceiptRenderer } from "../../renderers/receipt/esc-pos-receipt-renderer";
import { SimpleLabelRenderer } from "../../renderers/label/simple-label-renderer";
import {
  requireConnectedPrinter,
  validateLabelPayload,
  validateReceiptPayload
} from "../shared/adapter-validators";

const androidCapabilities: PrinterCapability = {
  ...getFallbackAndroidPrinterCapabilities(),
  labels: true,
  supportsCashDrawer: true,
  supportedTransports: ["usb", "bluetooth", "wifi", "socket"],
  supportedDocumentTypes: ["receipt", "label", "kot", "token", "retail-document"],
  supportedLabelSizesMm: ["102x40", "82x25"]
};

export class AndroidPrinterAdapter implements PrinterAdapter {
  private readonly nativeModule: AndroidNativePrinterModule;
  private readonly receiptRenderer = new EscPosReceiptRenderer();
  private readonly labelRenderer = new SimpleLabelRenderer();
  private connectedPrinter: PrinterDevice | null = null;

  constructor(nativeModule: AndroidNativePrinterModule = getAndroidNativePrinterModule()) {
    this.nativeModule = nativeModule;
  }

  async initialize(): Promise<void> {
    await this.callNative(() => this.nativeModule.initialize());
  }

  async discoverPrinters(): Promise<PrinterDevice[]> {
    const printers = await this.callNative(() => this.nativeModule.discoverPrinters());
    return printers.map((printer) => this.mapPrinter(printer));
  }

  async connect(printerId: string): Promise<PrinterDevice> {
    const printer = await this.callNative(() => this.nativeModule.connect(printerId));
    this.connectedPrinter = this.mapPrinter(printer);
    return this.connectedPrinter;
  }

  async disconnect(printerId?: string): Promise<void> {
    await this.callNative(() => this.nativeModule.disconnect(printerId));
    this.connectedPrinter = null;
  }

  async getConnectedPrinter(): Promise<PrinterDevice | null> {
    const printer = await this.callNative(() => this.nativeModule.getConnectedPrinter());
    this.connectedPrinter = printer ? this.mapPrinter(printer) : null;
    return this.connectedPrinter;
  }

  async getStatus(printerId?: string): Promise<PrinterStatus> {
    return this.callNative(() => this.nativeModule.getStatus(printerId));
  }

  async printReceipt(job: PrintJob, payload: ReceiptPayload): Promise<PrintResult> {
    validateReceiptPayload(payload, androidCapabilities);
    const printer = await this.getRequiredPrinter(job.targetPrinterId);
    const renderedChunks = await this.receiptRenderer.render(payload);
    const chunks = renderedChunks.map((chunk) => Array.from(chunk));
    const result = await this.callNative(() =>
      this.nativeModule.printReceipt({
        jobId: job.id,
        printerId: printer.id,
        chunks
      })
    );

    return {
      jobId: job.id,
      printer,
      status: result.status,
      message: result.message
    };
  }

  async printLabel(job: PrintJob, payload: LabelPayload): Promise<PrintResult> {
    validateLabelPayload(payload, androidCapabilities);
    const printer = await this.getRequiredPrinter(job.targetPrinterId);
    const renderedChunks = await this.labelRenderer.render(payload);
    const chunks = renderedChunks.map((chunk) => Array.from(chunk));
    const result = await this.callNative(() =>
      this.nativeModule.printLabel({
        jobId: job.id,
        printerId: printer.id,
        chunks
      })
    );

    return {
      jobId: job.id,
      printer,
      status: result.status,
      message: result.message
    };
  }

  async getCapabilities(query: AdapterCapabilityQuery): Promise<PrinterCapability[]> {
    if (query.platform !== "android") {
      return [];
    }

    if (query.transportTypes?.length) {
      const supportsRequestedTransport = query.transportTypes.some((transportType) =>
        androidCapabilities.supportedTransports.includes(transportType)
      );

      if (!supportsRequestedTransport) {
        return [];
      }
    }

    if (
      query.documentType &&
      !androidCapabilities.supportedDocumentTypes.includes(query.documentType)
    ) {
      return [];
    }

    return [androidCapabilities];
  }

  private async getRequiredPrinter(printerId: string): Promise<PrinterDevice> {
    const connectedPrinter = await this.getConnectedPrinter();
    return requireConnectedPrinter(connectedPrinter, printerId);
  }

  private mapPrinter(printer: NativePrinterDevice): PrinterDevice {
    return {
      id: printer.id,
      name: printer.name,
      platform: "android",
      transportType: printer.transportType,
      vendorId: printer.vendorId,
      productId: printer.productId,
      address: printer.address,
      capabilities: androidCapabilities
    };
  }

  private async callNative<T>(operation: () => Promise<T>): Promise<T> {
    try {
      return await operation();
    } catch (error) {
      if (error instanceof PrinterError) {
        throw error;
      }

      throw new PrinterError("NATIVE_ERROR", "Android native bridge call failed.", error);
    }
  }
}
